PostgreSQL support

abompard opened the pull request:
Comme indiqué dans ce post : http://www.bilboplanet.com/forum/viewtopic.php?pid=1241

Ces commits ajoutent le support de PostgreSQL dans Bilboplanet. Je ne pense pas qu'ils cassent le fonctionnement avec MySQL, mais je n'ai pas pu tester (je n'ai pas de MySQL ici).

Le commit 9ae1d9e touche aux types dans la base de données, je ne sais pas si tu as une procédure de migration de prévue pour ces cas-là...

Le commit 6b0b7ad n'a rien à voir avec PostgreSQL, désolé de le mettre dans le tas, mais bon, c'est un one-liner... De même pour le 37ca427

Voilà, c'est la première fois que je touche au code de Bilboplanet et je ne suis pas un expert en PHP, donc je suis très preneur de toutes tes remarques. Je suis prêt à retravailler le patchset autant que nécessaire.

abompard committed:
Fix the "LIMIT" SQL statement to be standards-compatible

The expression "LIMIT x,y" is MySQL-specific, replace it with LIMIT and
OFFSET statements.

abompard committed:
The SQL string-quoting char is the single quote

MySQL also accepts the double-quote to quote strings in SQL statements,
but PostgreSQL does not.

abompard committed:
Detect database size and version in PostgreSQL

The database size and server version detection code was only implemented
for MySQL. Add PostgreSQL support.

abompard committed:
Add ORDER BY criteria to the SELECT list

PostgreSQL requires all fields used as a sort criteria to be in the
SELECT list, even for a modified version (here: lower() applied)

abompard committed:
Fix database field types

- Text fields should not be limited in size
- 'post_id' is an integer, and should be declared as such in foreign keys
- 'tribe_search', 'tribe_tags', 'tribe_users' and 'tribe_icon' should have
  defaults, since there are requests in the code that inserts a line in their
  table without specifing them.

abompard committed:
Dividing by NULL in SQL is not allowed

Provide defaults when the "post" table is empty to avoid dividing by
NULL.

abompard committed:
PHP 5.4 compatibility

Call-time pass-by-reference has been removed.

abompard committed:
The 'user_id' field must not be NULL

The 'user_id' field must not be NULL when inserting in the 'post_tag'
table.

abompard committed:
Use "$core->con->escape()" instead of "addslashes()"

The "$core->con->escape()" method is backend-dependent, and thus better
suited to escape strings than the generic "addslashes()".

abompard committed:
Be stricter in SQL statements

PostgreSQL is stricter in the allowed SQL statements, fix some of them
for compatibility.

abompard committed:
Add a dummy file in admin/cache

Without this dummy file, git will not record the empty folder.

theclimber commented on inc/fonctions.php:
```diff
@@ -522,8 +522,11 @@ function generate_SQL(
+		$max_score = $max->f('max_score') or '1';
+		if (!$max_score) $max_score = '1';
+		$max_share = $max->f('max_share') or '1';
+		if (!$max_share) $max_share = '1';
```
Dans ce genre de cas tu peux utiliser une syntaxe plus compacte du type :
$max_view = !$max_view ? 1 : $max->f('max_view');

C'est en gros le "if" "then" "else" compact

abompard replied:
OK, mais ça n'économise pas de ligne ça, ou alors il faudrait un truc du genre :
    $max_view = !$max->f('max_view') ? 1 : $max->f('max_view');
Je vais faire ça.

theclimber replied:
Heu oui, t'as raison, au temps pour moi !! Je me suis trompé ... mais c'est sur le principe de gagner une ligne ;-)

theclimber commented on inc/lib/checkValidHTML.php:
```diff
@@ -76,7 +76,7 @@ class SafeHtmlChecker {
     function SafeHtmlChecker() {
         $this->parser = xml_parser_create();
-        xml_set_object($this->parser, &$this);
+        xml_set_object($this->parser, $this);
```
As-tu rencontré des erreurs avec le code de cette librairie? Comme elle n'est pas issue du bilboplanet ce n'est pas mon code. Mais je suis étonné car je n'ai pas eu d'erreurs jusqu'à maintenant

abompard replied:
Oui, sur PHP 5.4 ça ne compile pas du tout, j'ai l'erreur suivante :
PHP Fatal error:  Call-time pass-by-reference has been removed in [...]/inc/lib/checkValidHTML.php on line 82

theclimber replied:
J'avoue, j'ai pas encore testé le bilboplanet avec php 5.4 ... mais si c'est la seule erreur que tu as pour supporter le php 5.4 c'est plutôt une bonne nouvelle, ça veut dire que dans l'ensemble, le reste est supporté.

theclimber commented on admin/manage-feed.php:
```diff
@@ -106,7 +107,8 @@
 		'.$core->prefix.'user.user_id,
-		user_fullname
+		user_fullname,
+		lower(user_fullname) AS user_fullname_lower
```
Je vois que tu rajoutes à plusieurs endroits le "user_fullname_lower". Quel est la raison? J'ai l'impression que tu ne le ré-utilise pas ailleurs, donc je ne vois pas trop l'intérêt.

abompard replied:
C'est ce que j'ai mis dans le message de commit, mais effectivement c'est difficile à voir dans l'interface de GitHub (ou alors j'ai pas encore trouvé comment le montrer).
Dans une requête avec DISTINCT, PostgreSQL a besoin que tous les champs utilisés dans les critères de tri (ORDER BY) fassent partie des champs du SELECT, même si c'est une version modifiée du champ (ici, modifiée par la fonction lower()). MySQL fait le raccourci à ta place.
L'erreur retournée est "for SELECT DISTINCT, ORDER BY expressions must appear in select list".

theclimber replied:
Parfait, je comprends mieux. Pas de soucis à ce niveau là

theclimber commented on .gitignore:
```diff
@@ -2,6 +2,7 @@ inc/config.php
 inc/cron_running.txt
 inc/update.txt
 admin/cache/*
+!admin/cache/.keep
```
Pourquoi ce changement dans le gitignore?

abompard replied:
Quand j'ai fait l'extraction git, le répertoire admin/cache étant vide, il ne me l'a pas créé, j'ai donc eu des erreurs à la récupération des flux puisque le dossier n'existait pas. Une technique quand on veut garder un dossier vide est de mettre un fichier vide dedans, c'est ce que j'ai fait en ajoutant le fichier admin/cache/.keep. Ensuite, il faut bien sûr l'exclure du gitignore.

Si la méthode ne te plaît pas, je peux supprimer ce commit, qui n'a pas grand chose à voir avec PGSql...

theclimber replied:
Je ne connaissait pas cette technique. Mais tu peux la laisser. Effectivement, quand le répertoire cache n'existe pas, ça pose problème, donc laissons comme ça.

theclimber commented:
Encore un grand merci pour ton boulot, je t'ai répondu plus longueument sur le forum.
http://www.bilboplanet.com/forum/viewtopic.php?pid=1242#p1242
A+

abompard committed:
Small fixes

- Optimize 3 if statements (and win 3 lines ! Yay !)
- Fix 2 typos

abompard committed:
Use Clearbricks to export the database

Use Clearbricks instead of native MySQL functions.

Also ignore errors when unlinking a non-existent file.
